src/Frame.cpp

```diff
@@ -269,17 +268,22 @@ FrameId Frame::CreateCameraFrame(FrameId parent)
 
 void Frame::DeleteCameraFrame(FrameId camera)
 {
+	if (!camera)
+		return;
+
 	// Detach camera from parent, then delete:
 	Frame *cameraFrame = Frame::GetFrame(camera);
 	Frame *parent = Frame::GetFrame(cameraFrame->GetParent());
-	parent->RemoveChild(camera);
-	// Call dtor "popping" element in vector
-	#ifndef NDEBUG
-		if (camera != s_frames.size()) {
-			Error("DeleteCameraFrame: seems camera frame is not the last frame!\n");
-			abort();
-		};
-	#endif // NDEBUG
+	if (parent)
+		parent->RemoveChild(camera);
+
+// Call dtor "popping" element in vector
+#ifndef NDEBUG
+	if (camera != s_frames.size()) {
```

Comment:
I have to disclaim that this was me punching the code until it compiled and the game stopped crashing, and I have no idea what the broader context of these changes is, but shouldn't this be something like:
`if (camera != s_frames.back().GetId()) {`
Similarly at line 200:
`assert((s_frames.size() - 1) != f->m_thisId);`
should be:
`assert(s_frames.back().GetId() != f->m_thisId);`

Seems like otherwise the comparisons are to the size of the vector rather than the frameId?

In any case the game seems to work for me after I made these changes.

Reply:
> ...what the broader context...

Is a check made because `camera` is recreated every frame, thus I would check it was the last on that `vector`.

> Similarly at line 200:

This was to ensure save and load use the same ordering, as comment on previous line.

Probably I got a case of "off-by-one" error... But I don't know why I don't spot it before :/

Reply:
@vakhoir the FrameId as currently implemented is a raw index into the vector of Frame pointers. So to check if a FrameId is the last one in the list, a comparison of `camera == s_frames.size() - 1` would be correct; `s_frames.size()` is the index one-past-the-end of the vector. There's no need to dereference the pointer into the actual Frame to check if it's the last one.

That being said, I'll go over it with a fine-toothed comb to see if I can find off-by-one errors like in DeleteCameraFrame().

Reply:
> That being said, I'll go over it with a fine-toothed comb to see if I can find off-by-one errors like in DeleteCameraFrame().

@Web-eWorks You can narrow that search: it seems both are under debug clause (one in an `assert`-ion and one in `#ifndef NDEBUG`)...

Reply:
I just checked, and yeah, it's on off-by-one in line 282, the other line I quoted even has a -1 when comparing to the vector size. 

The DEBUG clause could explain another thing I was wondering about - how was this compiling for you guys. Because the comparison seems to be vector size to FrameId (as in: int to struct), rather than index to index. My compiler protested, anyway.

Reply:
> My compiler protested, anyway.

May I ask if with some warning or if it stop until you change something?

Reply:
```
pioneer/src/Frame.cpp:200:32: error: ambiguous overload for ‘operator!=’ (operand types are ‘std::vector<Frame>::size_type {aka long unsigned int}’ and ‘FrameId’)
   assert((s_frames.size() - 1) != f->m_thisId);
          ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
pioneer/src/Frame.cpp:200:32: note: candidate: operator!=(std::vector<Frame>::size_type {aka long unsigned int}, int) <built-in>
pioneer/src/Frame.cpp:200:32: note: candidate: operator!=(std::vector<Frame>::size_type {aka long unsigned int}, size_t {aka long unsigned int}) <built-in>
```

I missed the int() and size_t() cast operator overloads earlier, the code does make sense at fase value. It might be a gcc thing, or a question of some obscure settings.

Reply:
> It stopped until I explicitly called the .id() on the FrameIds

...which allow compiler to perform an **implicit** conversion outside FrameId thing... :D 

(or at least I guess so, I'm AFK right now...)

Anyway, W C++!!!

Reply:
I *could* have added an implicit conversion to `std::size_t`, but the problem with that is that invalid values would map onto the valid range by accident. It might work to `return m_id < 0 ? size_t(-1L) : m_id`, but it's easier to make the conversion explicit so that the programmer has to think about what they're doing.

Reply:
Well, I would say that I'm really happy when I see this kind of code. Personally I deferred it as I didn't know how much it will take to get it right so I give it up and use only "plain-old-basic functionality"...

That said, I would ask if you mind to open a dedicated PR only for the bugfix as it should merge sooner. You then gain time to refine this an have a glance on my experimental branch and include some (or all) the optimization I did there...

Briefly: I put 0 (zero) in FrameId as RootFrame and this let me decouple Space and lead to some more optimizations